llm/generate/gen_common.sh

```diff
@@ -39,8 +39,39 @@ init_vars() {
     *)
         ;;
     esac
-    if [ -z "${CMAKE_CUDA_ARCHITECTURES}" ] ; then 
-        CMAKE_CUDA_ARCHITECTURES="50;52;61;70;75;80"
+    if [ -f /etc/nv_tegra_release ] ; then
+        TEGRA_DEVICE=1
+        echo "Tegra device detected: ${JETSON_MODEL}, L4T Version ${JETSON_L4T}, Jetpack Version ${JETSON_JETPACK}"
+    fi
+    if [ -z "${CMAKE_CUDA_ARCHITECTURES}" ] ; then
+        if [ -n ${TEGRA_DEVICE} ]; then
+            echo "CMAKE_CUDA_ARCHITECTURES unset, values are needed for Tegra devices. Using default values defined at https://github.com/dusty-nv/jetson-containers/blob/master/jetson_containers/l4t_version.py"
+        fi
+        # Tegra devices will fail generate unless architectures are set:
+        # Nano/TX1 = 5.3, TX2 = 6.2, Xavier = 7.2, Orin = 8.7
+        # L4T_VERSION.major >= 36:    # JetPack 6
+        #     CUDA_ARCHITECTURES = [87]
+        # L4T_VERSION.major >= 34:  # JetPack 5
+        #     CUDA_ARCHITECTURES = [72, 87]
+        # L4T_VERSION.major == 32:  # JetPack 4
+        #     CUDA_ARCHITECTURES = [53, 62, 72]
+        case $(echo "${JETSON_JETPACK}" | cut -d"." -f1) in
+        "6")
+            echo "Jetpack 6 detected. Setting CMAKE_CUDA_ARCHITECTURES='87'"
+            CMAKE_CUDA_ARCHITECTURES="87"
+            ;;
+        "5")
+            echo "Jetpack 5 detected. Setting CMAKE_CUDA_ARCHITECTURES='72;87'"
+            CMAKE_CUDA_ARCHITECTURES="72;87"
+            ;;
+        "4")
+            echo "Jetpack 4 detected. Setting CMAKE_CUDA_ARCHITECTURES='53;62;72'"
+            CMAKE_CUDA_ARCHITECTURES="53;62;72"
+            ;;
```

Comment:
Is it not possible to build a single ARM binary that has architectures 53-87 so the same binary will work on the different jetpacks?

Reply:
I tried to compile it on my Jetson leaving it the default architectures and it failed to compile. I checked into dustynv's container for llama_cpp and he explicitly sets the architectures to these based on the version of L4T running. Once I did that, it compiled without issue. If you have an ARM build system that will incorporate the other changes in this PR but build for all the architectures, I can test if it works. 

I hadn't considered that this would create device-specific binaries (versus things that just load at runtime), I'll have to explore other options.